cpp/log/cert.cc

```diff
@@ -546,12 +566,314 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
       LOG(WARNING) << "Corrupt extension data";
       LOG_OPENSSL_ERRORS(WARNING);
     }
+
     return FALSE;
   }
 
   return TRUE;
 }
 
+bool Cert::IsRedactedHost(const std::string& hostname) {
+  // Split the hostname on '.' characters
+  std::vector<string> tokens = util::split(hostname, '.');
+
+  for (string str : tokens) {
+    if (str == "?") {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool Cert::IsValidRedactedHost(const std::string& hostname) {
+  // Split the hostname on '.' characters
+  std::vector<string> tokens = util::split(hostname, '.');
+
+  // Enforces the following rules: '?' must be to left of non redactions
+  // If first label is '*' then treat it as if it was a redaction
+  bool can_redact = true;
+  for (int pos = 0; pos < tokens.size(); ++pos) {
+    if (tokens[pos] == "?") {
+      if (!can_redact) {
+        return false;
+      }
+    } else {
+      // Allow a leading '*' for redaction but once we've seen anything else
+      // forbid further redactions
+      if (tokens[pos] != "*") {
+        can_redact = false;
+      } else if (pos > 0) {
+        // '*' is only valid at the left
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+Cert::Status Cert::IsValidWildcardRedaction() const {
+  if (!IsLoaded()) {
+    LOG(ERROR) << "Cert not loaded";
+    return Cert::ERROR;
+  }
+
+  std::string common_name;
+  std::vector<std::string> dns_alt_names;
+  int redacted_name_count = 0;
+
+  // First. Check all the Subject Alt Name extension records. Any that are of
+  // type DNS must pass validation if they are attempting to redact labels
+  STACK_OF(GENERAL_NAME) * subject_alt_names = (STACK_OF(GENERAL_NAME) *)
```

Comment:
You *may* be able to pull the `unique_ptr` with a custom deallocation functor trick here to easy the memory management effort, haven't tried it with a STACK_OF yet though, so YMMV.